src/java/org/apache/cassandra/service/ConsensusTableMigrationState.java

```diff
@@ -0,0 +1,697 @@
+            if (o == null || getClass() != o.getClass()) return false;
+            MigrationStateSnapshot that = (MigrationStateSnapshot) o;
+            return tableStates.equals(that.tableStates) && epoch.equals(that.epoch);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(tableStates, epoch);
+        }
+    }
+
+    private ConsensusTableMigrationState() {}
+
+    // Used by callers to avoid looking up the TMS multiple times
+    public static @Nullable TableMigrationState getTableMigrationState(long epoch, TableId tableId)
+    {
+        ClusterMetadata cm = ClusterMetadataService.instance.maybeCatchup(Epoch.create(0, epoch));
+        TableMigrationState tms = cm.migrationStateSnapshot.tableStates.get(tableId);
+        return tms;
+    }
+
+    /*
+     * Set or change the migration target for the keyspaces and tables. Can be used to reverse the direction of a migration
+     * or instantly migrate a table to a new protocol.
+     */
+    public static void setConsensusMigrationTargetProtocol(@Nonnull String targetProtocolName,
+                                                           @Nonnull List<String> keyspaceNames,
+                                                           @Nonnull Optional<List<String>> maybeTables)
+    {
+        checkArgument(!keyspaceNames.isEmpty(), "At least one keyspace must be specified");
+        checkArgument(keyspaceNames.size() == 1 || !maybeTables.isPresent(), "Can't specify tables with multiple keyspaces");
+        checkArgument(!maybeTables.isPresent() || !maybeTables.get().isEmpty(), "Must provide at least 1 table if Optional is not empty");
+        ConsensusMigrationTarget targetProtocol = ConsensusMigrationTarget.fromString(targetProtocolName);
+
+        if (DatabaseDescriptor.getLegacyPaxosStrategy() == LegacyPaxosStrategy.accord)
+            throw new IllegalStateException("Mixing a hard coded strategy with migration is unsupported");
+
+        if (!Paxos.useV2())
+            throw new IllegalStateException("Can't do any consensus migrations from/to PaxosV1, switch to V2 first");
+
+        List<TableId> tableIds = keyspacesAndTablesToTableIds(keyspaceNames, maybeTables);
+        ClusterMetadataService.instance.commit(new SetConsensusMigrationTargetProtocol(targetProtocol, tableIds));
+    }
+
+    public static void startMigrationToConsensusProtocol(@Nonnull String targetProtocolName,
+                                                         @Nonnull List<String> keyspaceNames,
+                                                         @Nonnull Optional<List<String>> maybeTables,
+                                                         @Nonnull Optional<String> maybeRangesStr)
+    {
+        checkArgument(!keyspaceNames.isEmpty(), "At least one keyspace must be specified");
+        checkArgument(keyspaceNames.size() == 1 || !maybeTables.isPresent(), "Can't specify tables with multiple keyspaces");
```

Comment:
could we interpret empty keyspaces as a command to start migration of all key spaces?

Reply:
Sure, changed it to add all the non-local key spaces if key spaces is empty or null.